cpp/src/arrow/array/util.cc

```diff
@@ -673,22 +673,14 @@ class RepeatedArrayFactory {
     std::shared_ptr<Buffer> values_buffer, offsets_buffer;
     auto size = static_cast<typename T::offset_type>(value->size());
 
-    typename T::offset_type length_casted;
-    if (length_ > std::numeric_limits<typename T::offset_type>::max()) {
-      return Status::Invalid(
-          "length exceeds the maximum value of offset_type: ", std::to_string(length_),
-          " is greater than ",
-          std::to_string(std::numeric_limits<typename T::offset_type>::max()));
-    }
-    length_casted = static_cast<typename T::offset_type>(length_);
-
     typename T::offset_type total_size;
-    if (MultiplyWithOverflow(size, length_casted, &total_size)) {
+    if (MultiplyWithOverflow(size, static_cast<typename T::offset_type>(length_),
```

Comment:
I think in order to make this correct, this multiplication must be moved into CreateOffsetsBuffer.